File: prototype/rails/src/app/domain/cart/cart_line.rb
require_relative "../money"

module Domain
  module Cart
    CartLine = Data.define(:seller_id, :unit_price, :quantity) do
      def initialize(seller_id:, unit_price:, quantity:)
        raise ArgumentError, "a cart line covers at least one item, got #{quantity}" if quantity < 1

        super
      end

      def total
        unit_price * quantity
      end
    end
  end
end

File: prototype/node/src/app/core/cart/cart-line.ts
import type { Cents } from '../money.ts'
import { multiplyCents } from '../money.ts'
import type { SellerId } from '../ids/entity-ids.ts'

export type CartLine = {
  sellerId: SellerId
  unitPriceCents: Cents
  quantity: number
}

export function createCartLine(line: CartLine): CartLine {
  if (line.quantity < 1) {
    throw new RangeError(`a cart line covers at least one item, got ${line.quantity}`)
  }
  return line
}

export function cartLineTotal(line: CartLine): Cents {
  return multiplyCents(line.unitPriceCents, line.quantity)
}
